### PR TITLE
Enable 'strictNullChecks' ts compiler option.

### PR DESCRIPTION
**Pull request by rsgowman**
In the majority of cases, I simply turned stuff like this:

```
function fname(param: T) {
  if (!param) throw Error("invalid param; must not be nullish");
  ...
}
```

into this:

```
function fname(param: T|null|undefined) {
  if (!param) throw Error("invalid param; must not be nullish");
  ...
}
```

A more sensible approach would be eliminate the possibility of
null/undef and eliminate the checks too, but that's a bit more involved
and this patch is already enormous. But I think these sorts of
improvements could be taken advantage of opportunistically as the code
is worked on in the future.

### CODE REVIEW

File: src/firestore/firestore.ts [hunk 1]
```diff
@@ -71,8 +71,11 @@ export function getFirestoreOptions(app: FirebaseApp): Settings {
     });
   }
 
-  const projectId: string = utils.getProjectId(app);
-  const cert: Certificate = tryGetCertificate(app.options.credential);
+  const projectId: string | null = utils.getProjectId(app);
+  let cert: Certificate | null = null;
+  if (app.options.credential) {
+    cert = tryGetCertificate(app.options.credential);
```
Comment (hiranya911) [inline src/firestore/firestore.ts]: Does it make sense to declare this method as:

```
tryGetCertificate(credential: Credential | null | undefined): Certificate | null;
```

File: src/messaging/messaging-errors.ts [hunk 2]
```diff
@@ -67,21 +67,22 @@ export function createFirebaseError(err: HttpError): FirebaseMessagingError {
  */
 export function getErrorCode(response: any): string | null {
   if (validator.isNonNullObject(response) && 'error' in response) {
-    if (validator.isString(response.error)) {
-      return response.error;
+    const error = (response as any).error;
```
Comment (hiranya911) [inline src/messaging/messaging-errors.ts]: Is this because `isNoneNullObject()` asserts the type of `response` to be something other than `any`?

File: src/utils/api-request.ts [hunk 3]
```diff
@@ -459,6 +470,11 @@ class AsyncHttpCall {
     try {
       this.config = new HttpRequestConfigImpl(config);
       this.options = this.config.buildRequestOptions();
+      if (!this.options.headers) {
```
Comment (hiranya911) [inline src/utils/api-request.ts]: This doesn't look right. It is possible for this to be undefined. May be initialize to empty object?

File: test/integration/auth.spec.ts [hunk 4]
```diff
@@ -246,12 +248,13 @@ describe('admin.auth', () => {
       })
       .then(() => {
         // New sign-in should succeed.
-        return firebase.auth().signInWithEmailAndPassword(
+        return firebase.auth!().signInWithEmailAndPassword(
```
Comment (hiranya911) [inline test/integration/auth.spec.ts]: Why is this needed?

File: test/integration/auth.spec.ts [hunk 5]
```diff
@@ -1384,11 +1409,16 @@ describe('admin.auth', () => {
         } as any,
         computePasswordHash: (userImportTest: UserImportTest): Buffer => {
           const currentRawPassword = userImportTest.rawPassword;
-          const currentRawSalt = userImportTest.rawSalt;
-          const N = userImportTest.importOptions.hash.memoryCost;
-          const r = userImportTest.importOptions.hash.blockSize;
-          const p = userImportTest.importOptions.hash.parallelization;
-          const dkLen = userImportTest.importOptions.hash.derivedKeyLength;
+          expect(userImportTest.rawSalt);
+          const currentRawSalt = userImportTest.rawSalt!;
```
Comment (hiranya911) [inline test/integration/auth.spec.ts]: Nit: Please add some whitespace after each `const` for clarity.

File: test/unit/auth/auth.spec.ts [hunk 6]
```diff
@@ -1236,7 +1242,7 @@ AUTH_CONFIGS.forEach((testConfig) => {
       });
 
       it('should be rejected given invalid properties', () => {
-        return auth.createUser(null)
+        return auth.createUser(null as unknown as CreateRequest)
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: Can we `null as any`?

File: test/unit/auth/auth.spec.ts [hunk 7]
```diff
@@ -411,6 +411,9 @@ AUTH_CONFIGS.forEach((testConfig) => {
       const tenantId = testConfig.supportsTenantManagement ? undefined : TENANT_ID;
       const expectedUserRecord = getValidUserRecord(getValidGetAccountInfoResponse(tenantId));
       // Set auth_time of token to expected user's tokensValidAfterTime.
+      if (!expectedUserRecord.tokensValidAfterTime) {
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: Can we use `!` assertion here and remove this `if` condition?

File: test/unit/auth/tenant-manager.spec.ts [hunk 8]
```diff
@@ -401,7 +401,7 @@ describe('TenantManager', () => {
     });
 
     it('should be rejected given invalid TenantOptions', () => {
-      return tenantManager.createTenant(null)
+      return tenantManager.createTenant(null as unknown as TenantOptions)
```
Comment (hiranya911) [inline test/unit/auth/tenant-manager.spec.ts]: `null as any` if that's allowed

File: test/unit/messaging/messaging.spec.ts [hunk 9]
```diff
@@ -1116,7 +1116,7 @@ describe('Messaging', () => {
 
     it('should throw given no registration token(s) argument', () => {
       expect(() => {
-        messaging.sendToDevice(undefined as string, mocks.messaging.payloadDataOnly);
+        messaging.sendToDevice(undefined as unknown as string, mocks.messaging.payloadDataOnly);
```
Comment (hiranya911) [inline test/unit/messaging/messaging.spec.ts]: `as any`

File: src/auth/auth-api-request.ts [hunk 10]
```diff
@@ -683,7 +683,7 @@ const LIST_INBOUND_SAML_CONFIGS = new ApiSettings('/inboundSamlConfigs', 'GET')
  * Class that provides the mechanism to send requests to the Firebase Auth backend endpoints.
  */
 export abstract class AbstractAuthRequestHandler {
-  protected readonly projectId: string;
+  protected readonly projectId: string | null;
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/auth/auth-api-request.ts]: There are some `admin.auth()` APIs that doesn't require project ID (custom token creation for example). Option (b) with deferred validation sounds like the right thing to do. We do a similar thing in `FirebaseTokenVerifier`. But that may be a larger refactor than we wish to do right now. I'm ok with keeping this as is, and changing it later.

File: test/unit/auth/auth.spec.ts [hunk 11]
```diff
@@ -2169,8 +2170,9 @@ AUTH_CONFIGS.forEach((testConfig) => {
 
         if (emailActionFlow.requiresSettings) {
           it('should reject when called without actionCodeSettings', () => {
-            return (auth as any)[emailActionFlow.api](email, undefined)
-              .should.eventually.be.rejected.and.have.property('code', 'auth/argument-error');
+            expect(() => {
+              (auth as any)[emailActionFlow.api](email, undefined);
+            }).to.throw(FirebaseAuthError).with.property('code', 'auth/argument-error');
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: This looks like a breaking change. Instead of rejecting the API now throws. We probably don't want that.

File: src/utils/api-request.ts (hunk 3 above)
Comment (hiranya911) [reply src/utils/api-request.ts]: I see. The method `buildEntity()` also has code that modifies headers in-place.  I think we can just do this:

```
this.entity = this.config.buildEntity(this.options.headers!);
```

If `buildRequestOptions` did not set headers, this will trigger a runtime error in `buildEntity()` when it attempts to modify headers. That sounds like an ok position for us to be in. WDYT?

File: test/unit/auth/auth.spec.ts (hunk 6 above)
Comment (hiranya911) [reply test/unit/auth/auth.spec.ts]: I have a slight preference towards `as any` since it's less things to type. But this ok too.

File: test/unit/auth/auth.spec.ts (hunk 11 above)
Comment (rsgowman) [reply test/unit/auth/auth.spec.ts]: I wouldn't call it a breaking change; I'd call it a bug fix. :)  If you violate the api contract, you expect it to fail. (See also https://g3doc.corp.google.com/firebase/jscore/g3doc/contributing/styleguide.md#public-apis-should-sanitize-their-inputs-aggressively which I'm definitely in favour of.)

But you're right that this is a behaviour change, and I'm trying to minimize those in this PR. I've reverted this for now, but I think this is worth discussing:
a) whether we should change this throughout (though obviously not during this PR)
b) if so, whether we consider it a breaking change

File: test/unit/auth/auth.spec.ts [hunk 12]
```diff
@@ -1243,7 +1243,7 @@ AUTH_CONFIGS.forEach((testConfig) => {
       });
 
       it('should be rejected given invalid properties', () => {
-        return auth.createUser(null as unknown as CreateRequest)
+        return auth.createUser(null as any as CreateRequest)
```
Comment (hiranya911) [inline test/unit/auth/auth.spec.ts]: `null as any` should be sufficient here right?

File: test/unit/auth/tenant-manager.spec.ts [hunk 13]
```diff
@@ -401,7 +401,7 @@ describe('TenantManager', () => {
     });
 
     it('should be rejected given invalid TenantOptions', () => {
-      return tenantManager.createTenant(null as unknown as TenantOptions)
+      return tenantManager.createTenant(null as any as TenantOptions)
```
Comment (hiranya911) [inline test/unit/auth/tenant-manager.spec.ts]: Again `null as any` should be sufficient?

File: test/unit/messaging/messaging.spec.ts [hunk 14]
```diff
@@ -398,10 +398,10 @@ describe('Messaging', () => {
   describe('send()', () => {
     it('should throw given no message', () => {
       expect(() => {
-        messaging.send(undefined as unknown as Message);
+        messaging.send(undefined as any as Message);
```
Comment (hiranya911) [inline test/unit/messaging/messaging.spec.ts]: I'd expect just `undefined as any` to be enough here.
